### ISSUE TITLE
Support for custom annotations

### ISSUE THREAD

**Issue by nixxxon** [flume/enthistory#39]
**Is your feature request related to a problem? Please describe.**
When using enthistory together with entgql it would be nice to be able to skip schema/type generation for history tables by supplying the `entgql.Skip()`-annotation,

**Describe the solution you'd like**
To be able to specify custom annotations/policies (mixins?) that should be added to the history schemas. Maybe `enthistory.Annotations` can be used for that?

**Comment by frisbm** [comment]
@nixxxon thanks again for submitting an issue, I was actually able to handle this by inheriting annotations from the original schema onto the history schemas and looking at the graphql codegen, TestSkip and TestSkipHistory do not get gen'd

**Comment by nixxxon** [comment]
Thank you @frisbm ! That seems like a good start but not exactly what I need.

I need the original schema to be generated but not the history schema. I also have a case where I need to add policies to the history schema only.

I think that a custom mixin (instead of using the originals mixin) would solve both of these issues?

Would be great if one could specify custom mixin(s), per schema, for example using the `enthistory.Annotations`.

**Comment by frisbm** [comment]
@nixxxon ah sorry, i misunderstood the original enhancement description. For the first case, that is actually currently handled via an enthistory annotation. You can use the exclude to skip history schema generation altogether

```go
func (TestExclude) Annotations() []schema.Annotation {
	return []schema.Annotation{
		enthistory.Annotations{
			Exclude: true,
		},
	}
}
```

For the second case, i'll try to get that together, I like the idea of adding that to the annotations as well, might be a bit difficult to represent it in the generated history schema but I think if i need to I can have it be computed instead of templated

**Comment by frisbm** [comment]
@nixxxon gotcha, i think i'm following now. You want the ability to set an annotation like `entgql.Skip()` on the original schema but have it only apply to the history schema. Totally understandable, should be able to accomplish that with the same feature for the policies/interceptors feature above

**Comment by frisbm** [comment]
@nixxxon not great news, unfortunately with the way enthistory is now both of those are actually looking impossible to accomplish. Looking at the way ent actually does the schema migrations, i'll probably need to move enthistory to something similar to it. The annotations approach didnt work because ent marshals the annotations to json and i was trying to allow users to set the policy, interceptors, and hooks there but since they are funcs it was rejecting the generation. As for the entgql skip on just the history, its a similar story, by the time the annotation makes its way to the generator its a map and I cant pull the necessary info from it like id be able to if it was still of the type `schema.Annotation`.

As for your situation, you can probably just eject your original and history schema from tracking to customize it how you want it for now.

On the original schema if you add the enthistory `Exclude: true` annotation and on the history model remove the `History: true` annotation, you will be able to customize the history model without fear of it being overwritten. Its gonna probably be a little while until I get around to rewriting the enthistory schema loading, maybe a couple of weeks

**Comment by nixxxon** [comment]
Thanks for looking into it @frisbm. If I set `History:false` the hooks won't work right?

Would it be possible to define a list of mixins that are merged with the `original{}.mixin()` using templatating, like:
```go
enthistory.Annotations{
	Mixins: []ent.Mixin{},
},
```
And:
```go
func (TodoHistory) Mixin() []ent.Mixin {
	return slices.Concat(Todo{}.Mixin(), CustomMixins)
}
```

**Comment by frisbm** [comment]
@nixxxon I think the only way that might work right now is with templating like you mentioned but even then im not sure, mixins are especially tough because ent doesn't track them directly on the load.Schema, it parses out their component parts and appends them where necessary

```go
type Schema struct {
	Name         string         `json:"name,omitempty"`
	Config       ent.Config     `json:"config,omitempty"`
	Edges        []*Edge        `json:"edges,omitempty"`
	Fields       []*Field       `json:"fields,omitempty"`
	Indexes      []*Index       `json:"indexes,omitempty"`
	Hooks        []*Position    `json:"hooks,omitempty"`
	Interceptors []*Position    `json:"interceptors,omitempty"`
	Policy       []*Position    `json:"policy,omitempty"`
	Annotations  map[string]any `json:"annotations,omitempty"`
}
```
 Because of that, its hard to inject the graph with the correct values, especially so when they are being injected into the schema model after the graph loading is complete. Im thinking what needs to happen here is a a two step process. The history schemas need to be generated first, before ent loads the graph so that changes like that can be represented without trying to bend ent into the right shape.

My idea for it would be to have a history gen run before the ent gen that has a similar setup to the entc package, and the extension is still used in the entc call but only for generating the history hooks and clients without needing to touch the graph

```go
package ent

//go:generate go run enthistory.go
//go:generate go run entc.go
```

**Comment by nixxxon** [comment]
I understand, thanks again @frisbm. 

For now I will move our privacy rules to the originals mixins so that the history schemas gets the same permissions at least.

**Comment by nixxxon** [comment]
I just realized that the history schemas are not added to the resolver (probably part of what you said earlier), only the types gets exposed through the api, which is better than nothing. 

I still think that it would be really nice to be able to control which history schemas should be a part of the api and have types, queries, mutations and resolvers generated only for those.

**Comment by frisbm** [comment]
@nixxxon yep, ill be trying to get that sorted out sooner than later because I have a pretty clear image of the direction to go for handling per-schema history configuration. Ill be out for a lot of next week but i'll try to aim for under two weeks total

**Comment by frisbm** [comment]
@nixxxon check out the latest release, v0.13.0, i added the ability to add custom annotations and mixins for history schemas

**Comment by nixxxon** [comment]
Thanks @frisbm  !
